keps/sig-node/5304-dra-attributes-downward-api/README.md

```diff
@@ -0,0 +1,969 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes exposing Dynamic Resource Allocation (DRA) device attributes to workloads via CDI (Container Device
+Interface) mounts. The DRA framework will provide helper functions enabling drivers to automatically generate per-claim
+attribute JSON files and mount them into containers via CDI. Workloads like KubeVirt can read device metadata like
+PCIe bus address, or mediated device UUID, from standardized file paths without requiring custom controllers or Downward
+API changes.
+
+## Motivation
+
+Workloads that need to interact with DRA-allocated devices (like KubeVirt virtual machines) require access to device
+specific metadata such as PCIe bus addresses or mediated device UUIDs. Currently, to fetch attributes from allocated
+devices, users must:
+1. Go to `ResourceClaimStatus` to find the request and device name
+2. Look up the `ResourceSlice` with the device name to get attribute values
+
+This complexity forces ecosystem projects like KubeVirt to build custom controllers that watch these objects and inject
+attributes via annotations/labels, leading to fragile, error-prone, and racy designs.
+
+### Goals
+
+- Provide a mechanism for workloads to discover DRA device metadata within workload pods.
+- Minimize complexity and avoid modifications to core components like scheduler and kubelet to maintain system
+  reliability and scalability
+- Provide an easy way for DRA device authors to make the attributes discoverable inside the pod.
+- Maintain full backward compatibility with existing DRA drivers and workloads
+
+### Non-Goals
+
+- Require changes to existing DRA driver implementations
+- Expose the entirety of `ResourceClaim`/`ResourceSlice` objects
+- Support dynamic updates to attributes after container start
+- Standardize attribute names or JSON schema in Alpha
+
+## Proposal
+
+This proposal introduces **framework-managed attribute JSON generation and CDI mounting** in the DRA kubelet plugin
+framework (`k8s.io/dynamic-resource-allocation/kubeletplugin`). Drivers opt-in by setting the `AttributesJSON(true)`
+option when starting their plugin.
+
+When enabled, the framework automatically:
+1. Generates a JSON file per claim+request containing device attributes
+2. Creates a corresponding CDI spec that mounts the attributes file into containers
+3. Appends the CDI device ID to the NodePrepareResources response
+4. Cleans up files during NodeUnprepareResources
+
+The workload reads attributes from the standardized path: `/var/run/dra-device-attributes/{driverName}-{claimUID}-{requestName}.json`
```

Comment:
how the application will know which `claimUID` is related to which request?
the pod can have multiple claims, inside it can have multiple requests, and a request can be for multiple devices

Reply:
what I am planing to have in the (https://github.com/k8snetworkplumbingwg/dra-driver-sriov) is `{driverName}-{resource-claim(from the pod)}-{request}.json`

example:

pod:
```
resourceClaims:
  - name: sriov
    resourceClaimTemplateName: single-vf
```

resourceClaim:
```
devices:
    requests:
    - exactly:
        allocationMode: ExactCount
        count: 1
        deviceClassName: sriovnetwork.k8snetworkplumbingwg.io
      name: vf
```

file name:
```
sriov-dra-sriov-vf.json
```

But in this case the json is a list as we can have multiple devices with the count